### ISSUE TITLE
Add support for propagating custom (e.g. zipkin) headers as span/trace id

### ISSUE THREAD

**Issue by skorhone** [elastic/apm-agent-java#716]
**Is your feature request related to a problem? Please describe.**
It's extremely hard to replace/utilize existing tracing infrastructure, when implementing new tracer would break the trace. Changing tracer for all infra structure components may take some (> months) time in a large organization

**Describe the solution you'd like**
Allow users to customize mapping for request headers using configuration.

http_headers:
  incoming:
    # Do not accept elastic-apm-traceparent
    elastic: false
   # Instead, read trace and span from custom headers (could be more than one provider)
    custom:
      traceId: ['X-B3-TraceId']
      spanId: ['X-B3-SpanId']
  outgoing:
    # Propagate headers to tracers that we have in our infrastructure (while each of microservices might be utilizing different trace)
    elastic: true
    custom:
      traceId: ['X-B3-TraceId']
      spanId: ['X-B3-SpanId']

It might make sense, to add support for specifying mapping per source / destination (ip / hostname).

**Describe alternatives you've considered**
None

**Additional context**
-

**Comment by felixbarny** [comment]
> It's extremely hard to replace/utilize existing tracing infrastructure, when implementing new tracer would break the trace

That's true and it's why we closely follow this spec: https://github.com/w3c/trace-context

We implement the current spec proposal with the difference that we currently call the header `elastic-apm-traceparent` as opposed to `traceparent`. But we intend to rename soon.

I'm not sure about the state of affairs in Zipkin but most likely they will eventually also support that header.

Unfortunately, that does not solve your immediate pain point but I'm not sure that we make a good time investment by supporting B3 headers if in the future all tracers will support the W3C standard.

Closing but happy to discuss further

**Comment by skorhone** [comment]
Even if zipkin (or name your tracing library here) adds support for w3c standard, one would have to redeploy all services. While I do get that use of standards is the way to go, the cost of migration as a big-bang project is hard to justify, if you already have a similar solution.

I'm not suggesting, that you should add support for specific provider. I rather thought, that it might be actually quite trivial to add a simple mapping algorithm. All alternatives that I can figure out, would add to infrastructure complexity. (We could implement mapping using reverse proxies or write custom agents).

**Comment by felixbarny** [comment]
Let's hear some more opinions from the team by transferring the issue to https://github.com/elastic/apm.
